internal/cmd/update.go

```diff
@@ -0,0 +1,466 @@
+	fmt.Fprintf(&b, "%d", version.Major())
+	fmt.Fprint(&b, ".")
+	fmt.Fprintf(&b, "%d", version.Minor())
+	fmt.Fprint(&b, ".")
+	fmt.Fprintf(&b, "%d", version.Patch())
+	if version.Prerelease() != "" {
+		fmt.Fprint(&b, "-")
+		fmt.Fprint(&b, version.Prerelease())
+	}
+
+	urlString := fmt.Sprintf("https://api.github.com/repos/cdr/coder-cli/releases/tags/v%s", b.String())
+	clog.LogInfo("query github releases", fmt.Sprintf("url: %q", urlString))
+
+	type asset struct {
+		BrowserDownloadURL string `json:"browser_download_url"`
+		Name               string `json:"name"`
+	}
+	type release struct {
+		Assets []asset `json:"assets"`
+	}
+	var r release
+
+	resp, err := httpClient.Get(urlString)
+	if err != nil {
+		return "", xerrors.Errorf("query github release url %s: %w", urlString, err)
+	}
+	defer resp.Body.Close()
+
+	err = json.NewDecoder(resp.Body).Decode(&r)
+	if err != nil {
+		return "", xerrors.Errorf("unmarshal github releases api response: %w", err)
+	}
+
+	var assetURLStr string
+	for _, a := range r.Assets {
+		if strings.HasPrefix(a.Name, "coder-cli-"+ostype) {
+			assetURLStr = a.BrowserDownloadURL
+		}
+	}
+
+	if assetURLStr == "" {
+		return "", xerrors.Errorf("could not find release for ostype %s", ostype)
+	}
+
+	return assetURLStr, nil
+}
+
+func extractFromArchive(path string, archive []byte) ([]byte, error) {
+	contentType := http.DetectContentType(archive)
+	switch contentType {
+	case "application/zip":
+		return extractFromZip(path, archive)
+	case "application/x-gzip":
+		return extractFromTgz(path, archive)
+	default:
+		return nil, xerrors.Errorf("unknown archive type: %s", contentType)
+	}
+}
+
+func extractFromZip(path string, archive []byte) ([]byte, error) {
+	zipReader, err := zip.NewReader(bytes.NewReader(archive), int64(len(archive)))
+	if err != nil {
+		return nil, xerrors.Errorf("failed to open zip archive")
+	}
+
+	var zf *zip.File
+	for _, f := range zipReader.File {
+		if f.Name == path {
+			zf = f
+			break
+		}
+	}
+	if zf == nil {
+		return nil, xerrors.Errorf("could not find path %q in zip archive", path)
+	}
+
+	rc, err := zf.Open()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to extract path %q from archive", path)
+	}
+	defer rc.Close()
+
+	var b bytes.Buffer
+	bw := bufio.NewWriter(&b)
+	if _, err := io.Copy(bw, rc); err != nil {
+		return nil, xerrors.Errorf("failed to copy path %q to from archive", path)
+	}
+	return b.Bytes(), nil
+}
+
+func extractFromTgz(path string, archive []byte) ([]byte, error) {
+	zr, err := gzip.NewReader(bytes.NewReader(archive))
+	if err != nil {
+		return nil, xerrors.Errorf("failed to gunzip archive")
+	}
+
+	tr := tar.NewReader(zr)
+
+	var b bytes.Buffer
+	bw := bufio.NewWriter(&b)
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, xerrors.Errorf("failed to read tar archive: %w", err)
+		}
+		fi := hdr.FileInfo()
+		if fi.Name() == path && fi.Mode().IsRegular() {
+			_, _ = io.Copy(bw, tr)
```

Comment:
seems like we should handle this error instead of ignoring it? I guess it could be that the tar we downloaded is corrupt?